### ISSUE TITLE
Improve structural navigation performance

### ISSUE THREAD

**Issue by TTWNO** [odilia-app/odilia#6]
Although there are some simpler issues to solve first, I want to leave this one up here for the future, when we have some more time to look at performance. The truth is, on a big page, it doesn't matter how fast the language you use is, if DBus is your bottleneck.
Granted, I'm glad that DBus can be our bottleneck. This is great news! But in order to increase the performance of hundreds, or on large pages, thousands of DBus calls all at once, we may need to consider alternative solutions to calling out to DBus every time we're checking a role.

Remember that the `zbus` crate already provide property caching, so things like the parent element, or the description of the Accessible are already cached for us. This is a massive improvement from calling out to DBus on literally every property or method comparison, but there are two key ways we may improve performance of these methods.

1. Local cache
2. Send MR to [atspi-atk](https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/85) for the Colleciton interface

## Local Cache

One option to increase performance on very large pages is to create a structure containing a few key pieces of information about each Accessible in the a11y hierarchy. 
It seems to me that a HashMap containing the accessible ID as the key, and the following struct for the value should be enough to carry performance from decent, to outstanding.

```rust
struct AccessibleCache {
  parent: i32;
  children: Vec<i32>;
  role: Role;
}
```

Of course, setting up a separate task to handle the enormous amount of created/destroyed/parent-changed/child-changed/text-changed/etc. events to keep a struct like this up to date, could indeed bring down the overall performance of Odilia in the process.
This is a performance hack, not a solution.

## Implementing Collection Support in ATK

I have [already submitted an issue](https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/85) to correct this problem in the a11y bridge library which Firefox uses.
Of course, the long term solution is to get this interface standardized and implemented at both the ATK (library) level, and Firefox (application level). The truth is that most applications do not need to implement this feature, since at most their hierarchy is going to be a few hundred a11y elements, maximum.
Web browsers are an extreme, outlier case, where you may have thousands of a11y elements in the tree of a single application.

Long term, this is what we should aim for. As of now, ATK does *not* implement the Collection interface at all.
The collection interface defines methods that help you perform structural navigation given 2 points and filter options.

1. The Collection to search (generally the root of the application, or the Pane as part of a web browser).
2. The Collection to start/stop at. So either a place to start from, or a place to stop at when performing a search. (This would generally be the currently focused element).
3. A set of binary OR'd match patterns, including Role and Attributes, as well as exactly how to match them: ANY, ALL, NONE, etc.

See here for the C implementation in `libatspi`: https://developer-old.gnome.org/libatspi/unstable/libatspi-atspi-collection.html

If this gets mainlined into major frameworks like GTK, Qt, and ATK (bridge), we will not only be making a more performant experience for Odilia, but for the entire accessibility stack. This is honestly something I should be looking into myself, but I'm putting this issue up for grabs if others want to take a stab at it.

**Comment by TTWNO** [comment]
Partially addressed by #9 , but it *feels* slower after this change.

**Comment by mcb2003** [comment]
Discussion of how we're going to do cache [starts here on Matrix](https://matrix.to/#/!BVPeWFRhpOwtjxKvET%3Amatrix.org/%24m4AUABIIEyTJaQwFVFyLwL9jk9WsLIQb_V38u88Og3s). Our design is now different: we're going to use multiple different data structures to speed up caching, including [`multimap`](https://crates.io/crates/multimap).

**Comment by TTWNO** [comment]
Proper caching needs to be implemented before 0.1.0, but I'll leace implementing the `Collection` interface for another time.

**Comment by albertotirla** [comment]
because cache events are now recieved, would it be safe to close this issue? I've seen that we're closer to 0.1.0, so I'm trying to see if anything can be closed as completed and wasn't because people forgot about it.

**Comment by TTWNO** [comment]
https://github.com/odilia-app/atspi/issues/49

**Comment by TTWNO** [comment]
Partially addressed by #2798484

**Comment by TTWNO** [comment]
Structural navigation performance should be good on a *desktop*, *x86* processor at this point, even if it's a bit older.
An upcoming PR should resolve issues on ARM Cortex A-72 range processors, whoes benchmarks currently look like this:

```text
cache/traverse_cache/zbus-docs
                        time:   [44.177 ms 44.182 ms 44.188 ms]
```
